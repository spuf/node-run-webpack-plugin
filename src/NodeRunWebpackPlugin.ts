import * as path from 'path';
import * as fs from 'fs';
import * as child_process from 'child_process';
import { ChildProcess } from 'child_process';
import { Plugin, Compiler, Stats } from 'webpack';

export interface Options {
  outputPath?: string;
  outputFilename?: string;
  context?: string;
  onChange?: (filepath: string) => void;
  onStart?: (filepath: string) => void;
  onStop?: () => void;
  onError?: (exitCode: number) => void;
}

interface OptionsWithDefaults extends Options {
  outputFilename: string;
  onChange: (filepath: string) => void;
  onStart: (filepath: string) => void;
  onStop: () => void;
  onError: (exitCode: number) => void;
}

const defaultOptions: OptionsWithDefaults = {
  outputFilename: 'index.js',
  onChange: filepath => {
    console.log(`# Changed ${filepath}, recompiling server...`);
  },
  onStart: filepath => {
    console.log(`# Run server ${filepath}`);
  },
  onStop: () => {
    console.log(`# Stopping server...`);
  },
  onError: exitCode => {
    console.log(`# Server exited with error, waiting for changes to restart...`);
  }
};

export default class NodeRunWebpackPlugin implements Plugin {
  private isWebpackWatching: boolean;
  private process: ChildProcess | null;
  private options: OptionsWithDefaults;

  constructor(options: Options = {}) {
    this.options = Object.assign({}, defaultOptions, options);
    this.isWebpackWatching = false;
    this.process = null;

    process.on('exit', () => {
      this.stopProcess();
    });
  }

  apply(compiler: Compiler) {
    this.updateOptions(compiler);

    compiler.plugin('watch-run', (_, callback: () => any) => {
      this.isWebpackWatching = true;
      callback();
    });

    compiler.plugin('watch-close', () => {
      this.isWebpackWatching = false;
    });

    compiler.plugin('invalid', (fileName: string) => {
      if (this.isWebpackWatching) {
        this.options.onChange(this.realtiveToContext(fileName));
        this.stopProcess();
      }
    });

    compiler.plugin('done', (stats: Stats) => {
      if (this.isWebpackWatching) {
        if (!stats.hasErrors()) {
          this.startProcess();
        }
      }
    });
  }

  private updateOptions(compiler: Compiler) {
    if (!this.options.outputPath) {
      if (!compiler.options.output || !compiler.options.output.path) {
        throw new Error('Undefined "output.path" webpack option.');
      }
      this.options.outputPath = compiler.options.output && compiler.options.output.path;
    }

    if (!this.options.context) {
      if (!compiler.options.context) {
        throw new Error('Undefined "context" webpack option.');
      }

      this.options.context = compiler.options.context;
    }
  }

  private realtiveToContext(filepath: string) {
    return this.options.context ? path.relative(this.options.context, filepath) : filepath;
  }

  private startProcess() {
    if (this.process) {
      return;
    }
    if (!this.options.outputPath || !this.options.outputFilename) {
      throw new Error('Undefined "outputPath" or "outputFilename" option.');
    }

    const nodeBin = process.argv[0];
    const outputPath = path.join(this.options.outputPath, this.options.outputFilename);

    this.options.onStart(this.realtiveToContext(outputPath));

    this.process = child_process.spawn(nodeBin, [outputPath], { stdio: 'inherit' });
    this.process.on('exit', code => {
      this.process = null;
      if (code) {
        this.options.onError(code);
      }
    });
  }

  private stopProcess() {
    if (this.process) {
      try {
        this.options.onStop();
      } finally {
        this.process.kill();
      }
    }
  }
}
