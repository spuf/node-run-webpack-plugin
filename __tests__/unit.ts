import { Options } from './../src/NodeRunWebpackPlugin';
import NodeRunWebpackPlugin from '../src/NodeRunWebpackPlugin';
import { Compiler, Stats } from 'webpack';

const CompilerMock = jest.fn<Compiler>((options, plugin) => ({ options, plugin }));

describe('constructor', () => {
  test('work without params', () => {
    const instance = new NodeRunWebpackPlugin();
    expect(instance).toBeInstanceOf(NodeRunWebpackPlugin);
    expect(instance).toMatchSnapshot();
  });
  test('work with params', () => {
    const instance = new NodeRunWebpackPlugin({ outputFilename: 'server.js' });
    expect(instance).toBeInstanceOf(NodeRunWebpackPlugin);
    expect(instance).toMatchSnapshot();
  });
  test('error about path', () => {
    const compiler = new CompilerMock({ output: {}, context: 'context' }, jest.fn());
    const instance = new NodeRunWebpackPlugin();
    expect(() => instance.apply(compiler)).toThrow('Undefined "output.path" webpack option.');
  });
  test('get path from options', () => {
    const compiler = new CompilerMock({ output: {}, context: 'context' }, jest.fn());
    const instance = new NodeRunWebpackPlugin({ outputPath: 'path' });
    expect(() => instance.apply(compiler)).not.toThrowError();
    expect(instance).toMatchSnapshot();
  });
  test('error about context', () => {
    const compiler = new CompilerMock({ output: { path: 'path' } }, jest.fn());
    const instance = new NodeRunWebpackPlugin();
    expect(() => instance.apply(compiler)).toThrow('Undefined "context" webpack option.');
  });
  test('get context from options', () => {
    const compiler = new CompilerMock({ output: { path: 'path' } }, jest.fn());
    const instance = new NodeRunWebpackPlugin({ context: 'context' });
    expect(() => instance.apply(compiler)).not.toThrowError();
    expect(instance).toMatchSnapshot();
  });
});
describe('apply', () => {
  const options: Options = {
    onStart: jest.fn(),
    onChange: jest.fn(),
    onError: jest.fn(),
    onStop: jest.fn()
  };
  const instance = new NodeRunWebpackPlugin(options);
  const eventMap: { [eventName: string]: Function } = {};
  const PluginMock = jest.fn<Function>((eventName: string, callback: Function) => {
    eventMap[eventName] = callback;
  });
  test('return void', () => {
    const compiler = new CompilerMock({ output: { path: 'path' }, context: 'context' }, PluginMock);
    expect(instance.apply(compiler)).toBe(undefined);
    expect(instance).toMatchSnapshot();
  });
  test('plugin events are set', () => {
    expect(PluginMock).toHaveBeenCalledTimes(4);
    expect(PluginMock).toHaveBeenCalledWith('watch-run', expect.any(Function));
    expect(PluginMock).toHaveBeenCalledWith('watch-close', expect.any(Function));
    expect(PluginMock).toHaveBeenCalledWith('invalid', expect.any(Function));
    expect(PluginMock).toHaveBeenCalledWith('done', expect.any(Function));
  });
  test('mock event map is populated', () => {
    expect(eventMap).toMatchObject({
      'watch-run': expect.any(Function),
      'watch-close': expect.any(Function),
      invalid: expect.any(Function),
      done: expect.any(Function)
    });
  });
  test('event done does nothing before watching', () => {
    const hasErrors = jest.fn();
    const StatsMock = jest.fn<Stats>(() => ({ hasErrors }));
    eventMap['done'](new StatsMock());
    expect(hasErrors).not.toHaveBeenCalled();
    expect(options.onStart).not.toHaveBeenCalled();
  });
  test('event invalid does nothing before watching', () => {
    eventMap['invalid']('file');
    expect(options.onChange).not.toHaveBeenCalled();
  });
  test('event watch-run is async', () => {
    const CallbackMock = jest.fn();
    eventMap['watch-run'](null, CallbackMock);
    expect(CallbackMock).toHaveBeenCalled();
  });
  test('event done does not start with errors', () => {
    const hasErrors = jest.fn(() => true);
    const StatsMock = jest.fn<Stats>(() => ({ hasErrors }));
    eventMap['done'](new StatsMock());
    expect(hasErrors).toHaveBeenCalledTimes(1);
    expect(options.onStart).not.toHaveBeenCalled();
  });
  test('event done starts without errors', () => {
    const hasErrors = jest.fn(() => false);
    const StatsMock = jest.fn<Stats>(() => ({ hasErrors }));
    eventMap['done'](new StatsMock());
    expect(hasErrors).toHaveBeenCalledTimes(1);
    expect(options.onStart).toHaveBeenCalledTimes(1);
    expect(options.onStart).toHaveBeenCalledWith('../path/index.js');
  });
  test('event invalid gives filename', () => {
    eventMap['invalid']('file');
    expect(options.onChange).toHaveBeenCalledTimes(1);
    expect(options.onChange).toHaveBeenCalledWith('../file');
  });
  test('event watch-close works', () => {
    expect(() => eventMap['watch-close']()).not.toThrowError();
  });
  test('lock event states', () => {
    expect(options.onStart).toHaveBeenCalledTimes(1);
    expect(options.onStop).toHaveBeenCalledTimes(1);
    expect(options.onChange).toHaveBeenCalledTimes(1);
    expect(options.onError).toHaveBeenCalledTimes(0);
  });
});
