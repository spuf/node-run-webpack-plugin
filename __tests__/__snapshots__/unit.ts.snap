// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply return void 1`] = `
NodeRunWebpackPlugin {
  "isWebpackWatching": false,
  "options": Object {
    "context": "context",
    "onChange": [Function],
    "onError": [Function],
    "onStart": [Function],
    "onStop": [Function],
    "outputFilename": "index.js",
    "outputPath": "path",
  },
  "process": null,
}
`;

exports[`constructor get context from options 1`] = `
NodeRunWebpackPlugin {
  "isWebpackWatching": false,
  "options": Object {
    "context": "context",
    "onChange": [Function],
    "onError": [Function],
    "onStart": [Function],
    "onStop": [Function],
    "outputFilename": "index.js",
    "outputPath": "path",
  },
  "process": null,
}
`;

exports[`constructor get path from options 1`] = `
NodeRunWebpackPlugin {
  "isWebpackWatching": false,
  "options": Object {
    "context": "context",
    "onChange": [Function],
    "onError": [Function],
    "onStart": [Function],
    "onStop": [Function],
    "outputFilename": "index.js",
    "outputPath": "path",
  },
  "process": null,
}
`;

exports[`constructor work with params 1`] = `
NodeRunWebpackPlugin {
  "isWebpackWatching": false,
  "options": Object {
    "onChange": [Function],
    "onError": [Function],
    "onStart": [Function],
    "onStop": [Function],
    "outputFilename": "server.js",
  },
  "process": null,
}
`;

exports[`constructor work without params 1`] = `
NodeRunWebpackPlugin {
  "isWebpackWatching": false,
  "options": Object {
    "onChange": [Function],
    "onError": [Function],
    "onStart": [Function],
    "onStop": [Function],
    "outputFilename": "index.js",
  },
  "process": null,
}
`;
